### PR TITLE
ARROW-5632: [Doc] Basic instructions for using Xcode with Arrow

### PR DESCRIPTION
**Pull request by hatemhelal**
Added a new section on debugging with Xcode on macOS.  I am a bit wary of doing a step-by-step tutorial with screenshots as those tend to get quickly out of date as the Xcode app evolves, and I think there will always be better sources elsewhere for helping users ramp up on how to use Xcode.  The goal is to add enough to help someone familiar with C++ development in Xcode to get started working on Arrow.  

cc: @nealrichardson or @xhochy: I recall you both have macs, could you try these instructions out and let me know if you run into any trouble?  Any advice on clarity, readability, etc would also be very welcome.  I'm not sure if I found the right location to add this so let me know if this fits in ok with the rest of the overall cpp doc.

### CODE REVIEW

**Comment by nealrichardson** [comment]
Sure, I can run the code example. What should I be looking for? I'm not familiar with the Xcode app and generally only open it by accident, so other than force-quitting, I don't have much experience with it ;)

**Comment by pprudhvi** [comment]
Fails to generate Xcode project file when Gandiva JNI bindings are enabled(-DARROW_GANDIVA_JAVA=ON). 

Errors:
------------------------------------------------------------------------
```
CMake Error in src/gandiva/jni/CMakeLists.txt:
  Target "gandiva_jni_shared" INTERFACE_INCLUDE_DIRECTORIES property contains
  path:

    "/Users/prudhvi/dev/arrow/cpp/xcode-build/src/gandiva/jni/java"

  which is prefixed in the build directory.


CMake Error in src/gandiva/jni/CMakeLists.txt:
  Target "gandiva_jni_shared" INTERFACE_INCLUDE_DIRECTORIES property contains
  path:

    "/Users/prudhvi/dev/arrow/cpp/xcode-build/src/gandiva/jni/java"

  which is prefixed in the build directory.Target "gandiva_jni_shared"
  INTERFACE_INCLUDE_DIRECTORIES property contains path:

    "/Users/prudhvi/dev/arrow/cpp/xcode-build/src/gandiva/jni/java"

  which is prefixed in the source directory.


CMake Error in src/gandiva/jni/CMakeLists.txt:
  Target "gandiva_jni_static" INTERFACE_INCLUDE_DIRECTORIES property contains
  path:

    "/Users/prudhvi/dev/arrow/cpp/xcode-build/src/gandiva/jni/java"

  which is prefixed in the build directory.


CMake Error in src/gandiva/jni/CMakeLists.txt:
  Target "gandiva_jni_static" INTERFACE_INCLUDE_DIRECTORIES property contains
  path:

    "/Users/prudhvi/dev/arrow/cpp/xcode-build/src/gandiva/jni/java"

  which is prefixed in the build directory.Target "gandiva_jni_static"
  INTERFACE_INCLUDE_DIRECTORIES property contains path:

    "/Users/prudhvi/dev/arrow/cpp/xcode-build/src/gandiva/jni/java"

  which is prefixed in the source directory.
```

**Comment by hatemhelal** [comment]
@nealrichardson, I think this can be tested at the command-line by generating the project with `cmake` followed by calling `xcodebuild` to confirm that it builds.  Try this with some different Arrow components enabled to flush out any latent problems.  Some things you could out try in Xcode if you are feeling adventurous:

* Select a unittest executable (e.g. `arrow-array-test`) to run from the scheme dropdown (top left, after the run and stop buttons). Cmd + R will run it.  
* Cmd + Shift + O will open a search dialog for jumping to a file or symbol.

@pprudhvi, I think this is a latent problem with the build setup for the Gandiva JNI bindings.  I opened a separate ticket to track this: [ARROW-5638](https://issues.apache.org/jira/browse/ARROW-5638)

**Comment by wesm** [comment]
@pprudhvi I took a brief look at this myself on macOS. It's a bug in the build system (there is already a hackaround in add_arrow_lib about Xcode CMAKE_GENERATOR), suggest it be debugged and fixed in ARROW-5638

+1

File: docs/source/developers/cpp.rst [hunk 1]
```diff
@@ -630,6 +630,27 @@ The report is then generated in ``compat_reports/libarrow`` as a HTML.
...
+
+Xcode is the IDE provided with macOS and can be use to develop and debug Arrow 
+by generating an Xcode project:
+
+.. code-block:: shell
+
+   cd cpp
+   mkdir xcode-build
+   cd xcode-build
+   cmake .. -G Xcode ^
```
Comment (kou) [inline docs/source/developers/cpp.rst]: We need to use "\\" for shell. "^" is for batch file.

Comment (hatemhelal) [reply docs/source/developers/cpp.rst]: Good catch @kou, I'll fix this in a followup PR.

**Pull request closed without merging**